### PR TITLE
Prepare for creating Dirac limits for prompt HNL analysis

### PR DESCRIPTION
**Pull request by lwezenbe**
### Changes

- Add functions to probe provenance of HNL leptons as much as possible
- Add functions to match l1 and SM neutrino in process
- Add switch to store whether an HNL event is Dirac type
- Add a hcounter for Dirac type events
- Add a function to see if event is Dirac type. It is based on the fact that the SM neutrino will be the opposite matter-antimatter type than l1 in the case of Dirac neutrinos where no LNV process is allowed and will also work for HNL decay to a SM neutrino and a Z

### Tests
Ran the tests without the switch on since they are not designed for the test files. All output remained the same as expected

### CODE REVIEW

File: multilep/interface/GenTools.h [hunk 1]
```diff
@@ -31,14 +31,17 @@ namespace GenTools{
...
     bool udsInChain(const std::set<int>&);
     bool tauInChain(const std::set<int>&);
+    bool heavyNeutrinoInChain(const std::set<int>&);
     //find the provenance of a particle using the contents of its decayChain
     unsigned provenance(const reco::GenParticle*, const std::vector<reco::GenParticle>&);
     unsigned provenanceCompressed(const reco::GenParticle*, const std::vector<reco::GenParticle>&, bool isPrompt);
 
     //check whether photon comes from ME in conversion
     unsigned provenanceConversion(const reco::GenParticle*, const std::vector<reco::GenParticle>&);
+    unsigned provenanceHeavyNeutrino(const reco::GenParticle&, const std::vector<reco::GenParticle>&);
```
Comment (knollejo) [inline multilep/interface/GenTools.h:41]: This should either be int, or the corresponding branch should be unsigned.

File: multilep/src/GenTools.cc [hunk 2]
```diff
@@ -213,12 +217,50 @@ unsigned GenTools::provenanceConversion(const reco::GenParticle* photon, const s
...
 }
 
+/*
+ * Check if the HNL sample is a Dirac type event or a LNV event
+ */
+bool GenTools::isDiracType(const std::vector<reco::GenParticle>& genParticles){
+    const reco::GenParticle* l1 = returnL1MatchForHNL(genParticles);
+    const reco::GenParticle* nu = returnNeutrinoMatchForHNL(genParticles);
+    if(l1->pdgId() > 0 && nu->pdgId() > 0) return false;
+    if(l1->pdgId() < 0 && nu->pdgId() < 0) return false;
```
Comment (knollejo) [inline multilep/src/GenTools.cc]: Can be simplified:

`return l1->pdgId() * nu->pdgId() < 0;`

Comment (lwezenbe) [reply multilep/src/GenTools.cc]: Done

File: multilep/interface/GenTools.h (hunk 1 above)
Comment (lwezenbe) [reply multilep/interface/GenTools.h:41]: The corresponding branch is now unsigned

**Review by knollejo — APPROVED** [review]

**Pull request merged**